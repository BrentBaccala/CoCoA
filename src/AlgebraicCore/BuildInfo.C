//   Copyright (c)  2005,2007,2012  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/config.H"

#include <limits>
#include <iostream>
using std::ostream;
using std::endl;
#include <string>
using std::string;

namespace CoCoA
{

  namespace BuildInfo
  {
    // Build info is actually passed in via three preprocessor variables:
    // COCOA_VERSION, COCOA_CXX, COCOA_CXXFLAGS -- all three are quoted strings.

    const std::string& version()
    {
      static const string info(COCOA_VERSION);
      return info;
    }

    const std::string& compiler()
    {
      static const string info(COCOA_CXX);
      return info;
    }

    const std::string& CompilationFlags()
    {
      static const string info(COCOA_CXXFLAGS);
      return info;
    }

    const std::string& CompilationDefines()
    {
      static const string info(COCOA_CXXFLAGS_DEFINES);
      return info;
    }


    void PrintAll(std::ostream& out)
    {
    // Below we use string literal juxtaposition, so that the build
    // info can readily be extracted from libcocoa.a by the command
    // strings libcocoa.a | egrep "CoCoA::BuildInfo"
      out << endl
          << "CoCoA::BuildInfo Summary of build information for CoCoALib:" << endl
          << "CoCoA::BuildInfo CoCoALib Version: " COCOA_VERSION << endl
          << "CoCoA::BuildInfo Compiler: " COCOA_CXX << endl
          << "CoCoA::BuildInfo Compilation Flags: " COCOA_CXXFLAGS << endl
          << "CoCoA::BuildInfo Compilation Flags Defines: " COCOA_CXXFLAGS_DEFINES << endl
          << "CoCoA::BuildInfo NumBits int = " << std::numeric_limits<unsigned int>::digits << endl
          << "CoCoA::BuildInfo NumBits long = " << std::numeric_limits<unsigned long>::digits << endl
          << "CoCoA::BuildInfo NumBits SmallExponent_t = " << std::numeric_limits<SmallExponent_t>::digits << endl
          << endl;
    }

  } // end of namespace CoCoA::BuildInfo

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/BuildInfo.C,v 1.8 2012/10/24 12:10:45 abbott Exp $
// $Log: BuildInfo.C,v $
// Revision 1.8  2012/10/24 12:10:45  abbott
// Revised PrintAll: now prints also NumBits of SmallExponent_t
//
// Revision 1.7  2012/10/02 11:45:09  bigatti
// -- added CompilationDefines
//
// Revision 1.6  2012/10/02 10:36:12  abbott
// Revised interface to BuildInfo information strings.
// Several consequential changes.
//
// Revision 1.5  2012/06/20 12:26:26  bigatti
// -- now exporting also compiler and compilation flags
//
// Revision 1.4  2011/03/11 17:38:21  bigatti
// -- added missing include
//
// Revision 1.3  2011/03/11 17:18:55  bigatti
// -- added info on int and long size
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.4  2007/02/10 18:44:03  cocoa
// Added "const" twice to each test and example.
// Eliminated dependency on io.H in several files.
// Improved BuildInfo, and added an example about how to use it.
// Some other minor cleaning.
//
// Revision 1.3  2007/02/08 22:34:22  cocoa
// Changed BuildInfo: only BuildInfo version is publicly visible.
// Only BuildInfo needs complicated compile-time flags, so several changes
// to Makefiles etc.  Added a new example: ex-BuildInfo.C
//
// Revision 1.1.1.1  2005/10/17 10:46:54  cocoa
// Imported files
//
// Revision 1.2  2005/10/12 15:52:09  cocoa
// Completed test-RingFp1 and corrected/cleaned the SmallFp*
// and RingFp* files.
//
// Some minor tidying elsewhere.
//
// Revision 1.1  2005/10/06 16:36:42  cocoa
// Added the capability find out build information at run-time.
// The Makefiles should be a little tidier too.
//
