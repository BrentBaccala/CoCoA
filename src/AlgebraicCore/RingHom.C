//   Copyright (c)  2005  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/RingHom.H"
#include "CoCoA/error.H"
#include "CoCoA/ideal.H"

#include <iostream>
using std::ostream;
#include<vector>
using std::vector; // used only in IsInKer

namespace CoCoA
{

  //-------------------------------------------------------
  // CLASS PartialRingHom

  RingElem PartialRingHom::operator()(ConstRefRingElem x) const
  {
    if (owner(x) != domain(*this))
      CoCoA_ERROR(ERR::BadRingHomArg, "Applying PartialRingHom to RingElem");
    RingElem ans(codomain(*this));
    mySmartPtr->myApply(raw(ans), raw(x));
    return ans;
  }


  RingElem PartialRingHom::operator()(const MachineInt& n) const
  {
    RingElem arg(domain(*this), n);
    RingElem ans(codomain(*this));
    mySmartPtr->myApply(raw(ans), raw(arg));
    return ans;
  }

  RingElem PartialRingHom::operator()(const BigInt& N) const
  {
    RingElem arg(domain(*this), N);
    RingElem ans(codomain(*this));
    mySmartPtr->myApply(raw(ans), raw(arg));
    return ans;
  }

  RingElem PartialRingHom::operator()(const BigRat& q) const
  {
    RingElem arg(domain(*this), q);
    RingElem ans(codomain(*this));
    mySmartPtr->myApply(raw(ans), raw(arg));
    return ans;
  }


  PartialRingHom PartialRingHom::operator()(const RingHom& theta) const
  {
    if (codomain(theta) != domain(*this))
      CoCoA_ERROR(ERR::BadCompose, "composition PartialRingHom(RingHom)");
// BUG BUG BUG can be clever here!!!
    return sequential(*this, theta);
  }


  PartialRingHom PartialRingHom::operator()(const PartialRingHom& theta) const
  {
    if (codomain(theta) != domain(*this))
      CoCoA_ERROR(ERR::BadCompose, "composition PartialRingHom(PartialRingHom)");
    return sequential(*this, theta);
  }


  void RingHomBase::myOutputSelf(std::ostream& out) const
  {
    out << "PartialRingHom(" << myDomain << " --> " << myCodomain;
    myOutputSelfDetails(out);
    out << ")";
  }


  void RingHomBase::myOutputSelfDetails(std::ostream& /*out*/) const
  {
    // Default definition does nothing (as there are no extra details to print).
    // SHOULD THIS BE PURE VIRTUAL (with no default defn)???
  }


  std::ostream& operator<<(std::ostream& out, const PartialRingHom& phi)
  {
    out << "RingHom(" << domain(phi) << " --> " << codomain(phi);
    phi->myOutputSelfDetails(out);
    out << ")";
    return out;
  }


  // better to use indices or iterators in this fn???
// USE AN ALGORITHM!!!
  bool IsInKer(const ideal& I, const PartialRingHom& phi)
  {
    const vector<RingElem>& g = gens(I);
    const vector<RingElem>::const_iterator stop = g.end();
    for (vector<RingElem>::const_iterator it = g.begin(); it != stop; ++it)
    {
      if (!IsZero(phi(*it))) return false;
    }
    return true;
  }


  //-------------------------------------------------------
  // CLASS RingHom

  RingElem RingHom::operator()(ConstRefRingElem x) const
  {
    if (owner(x) != domain(*this))
      CoCoA_ERROR(ERR::BadRingHomArg, "Applying RingHom to RingElem");
    RingElem ans(codomain(*this));
    mySmartPtr->myApply(raw(ans), raw(x));
    return ans;
  }


  RingElem RingHom::operator()(const MachineInt& n) const
  {
    RingElem arg(domain(*this), n);
    RingElem ans(codomain(*this));
    mySmartPtr->myApply(raw(ans), raw(arg));
    return ans;
  }

  RingElem RingHom::operator()(const BigInt& N) const
  {
    RingElem arg(domain(*this), N);
    RingElem ans(codomain(*this));
    mySmartPtr->myApply(raw(ans), raw(arg));
    return ans;
  }

  RingElem RingHom::operator()(const BigRat& q) const
  {
    RingElem arg(domain(*this), q);
    RingElem ans(codomain(*this));
    mySmartPtr->myApply(raw(ans), raw(arg));
    return ans;
  }


  RingHom RingHom::operator()(const RingHom& theta) const
  {
    if (codomain(theta) != domain(*this))
      CoCoA_ERROR(ERR::BadCompose, "composition  RingHom(RingHom)");
    return domain(theta)->myCompose(*this, theta);
  }


  PartialRingHom RingHom::operator()(const PartialRingHom& theta) const
  {
    if (codomain(theta) != domain(*this))
      CoCoA_ERROR(ERR::BadCompose, "composition  RingHom(PartialRingHom)");
    return sequential(*this, theta);
  }


//   void RingHomBase::myOutputSelf(std::ostream& out) const
//   {
//     out << "RingHom(" << myDomain << " --> " << myCodomain;
//     myOutputSelfDetails(out);
//     out << ")";
//   }


//   void RingHomBase::myOutputSelfDetails(std::ostream& /*out*/) const
//   {
//     // Default definition does nothing (as there are no extra details to print).
//     // SHOULD THIS BE PURE VIRTUAL (with no default defn)???
//   }


  std::ostream& operator<<(std::ostream& out, const RingHom& phi)
  {
    out << "RingHom(" << domain(phi) << " --> " << codomain(phi);
    phi->myOutputSelfDetails(out);
    out << ")";
    return out;
  }


  // better to use indices or iterators in this fn???
// USE AN ALGORITHM!!!
  bool IsInKer(const ideal& I, const RingHom& phi)
  {
    const vector<RingElem>& g = gens(I);
    const vector<RingElem>::const_iterator stop = g.end();
    for (vector<RingElem>::const_iterator it = g.begin(); it != stop; ++it)
    {
      if (!IsZero(phi(*it))) return false;
    }
    return true;
  }



  //---------------------------------------------------------------------------

  class IdentityRingHomImpl: public RingHomBase
  {
  private:
    explicit IdentityRingHomImpl(const ring& R);
    friend RingHom IdentityHom(const ring& R); // The only function that calls the ctor.
  public:
    virtual void myApply(RingElemRawPtr image, RingElemConstRawPtr arg) const;
    virtual bool IamPartial() const { return false; }
    virtual void myOutputSelfDetails(std::ostream& out) const;
  };


  IdentityRingHomImpl::IdentityRingHomImpl(const ring& R):
      RingHomBase(R, R)
  {}


  void IdentityRingHomImpl::myApply(RingElemRawPtr image, RingElemConstRawPtr arg) const
  {
    myDomain->myAssign(image, arg);
  }


  void IdentityRingHomImpl::myOutputSelfDetails(std::ostream& out) const
  {
    out << " the identity";
  }


  RingHom IdentityHom(const ring& R)
  {
    return RingHom(new IdentityRingHomImpl(R));
  }


  //---------------------------------------------------------------------------

  class SequentialRingHomImpl: public RingHomBase
  {
  private:
    SequentialRingHomImpl(const PartialRingHom& phi, const PartialRingHom& theta);
    friend PartialRingHom sequential(const PartialRingHom& phi, const PartialRingHom& theta); // The only function that calls the ctor.
  public:
    virtual void myApply(RingElemRawPtr image, RingElemConstRawPtr arg) const;
    virtual bool IamPartial() const { return IsPartial(myPhi) || IsPartial(myTheta); }
    virtual void myOutputSelfDetails(std::ostream& out) const;
  private: // data members
    PartialRingHom myPhi;
    PartialRingHom myTheta;
  };


  SequentialRingHomImpl::SequentialRingHomImpl(const PartialRingHom& phi, const PartialRingHom& theta):
      RingHomBase(domain(theta), domain(phi)),
      myPhi(phi),
      myTheta(theta)
  {
    CoCoA_ASSERT(domain(myPhi) == codomain(myPhi));
  }


  void SequentialRingHomImpl::myApply(RingElemRawPtr image, RingElemConstRawPtr arg) const
  {
//    RingElem tmp = myPhi(myTheta(arg));
    RingElem tmp = myPhi(myTheta(RingElemAlias(domain(myTheta),arg)));
    myDomain->myAssign(image, raw(tmp));
  }


  void SequentialRingHomImpl::myOutputSelfDetails(std::ostream& out) const
  {
    out << " composite of " << myPhi << " applied to " << myTheta;
  }


  PartialRingHom sequential(const PartialRingHom& phi, const PartialRingHom& theta)
  {
    if (domain(phi) != codomain(theta))
      CoCoA_ERROR(ERR::BadCompose, "sequential(phi,theta)");
    return PartialRingHom(new SequentialRingHomImpl(phi, theta));
  }


  bool ImageLiesInSubfield(const RingHom& phi)
  {
    if (IsField(domain(phi)) || IsField(codomain(phi))) return true;
    return codomain(phi)->myImageLiesInSubfield(phi);
  }


  //---------------------------------------------------------------------------

  void RingHomEmbeddingBase::myOutputSelfDetails(std::ostream& out) const
  {
    out << " canonical embedding";
  }

  //---------------------------------------------------------------------------

  RingHomInducedBase::RingHomInducedBase(const ring& NewDomain, const RingHom& InducingHom):
      RingHomBase(NewDomain, codomain(InducingHom)),
      myInducingHom(InducingHom)
  {}


  void RingHomInducedBase::myOutputSelfDetails(std::ostream& out) const
  {
    out << " induced by " << myInducingHom;
  }


} // end of namespace CoCoA


// RCS header/log
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/RingHom.C,v 1.8 2013/02/21 14:14:42 abbott Exp $
// $Log: RingHom.C,v $
// Revision 1.8  2013/02/21 14:14:42  abbott
// First attempt at implementing PartialRingHom -- some problems remain!!
//
// Revision 1.7  2011/11/09 14:11:58  bigatti
// -- renamed MachineInteger --> MachineInt
//
// Revision 1.6  2011/08/24 10:28:49  bigatti
// -- renamed QQ --> BigRat
// -- sorted #include
//
// Revision 1.5  2011/08/14 15:52:16  abbott
// Changed ZZ into BigInt (phase 1: just the library sources).
//
// Revision 1.4  2011/02/23 15:00:03  bigatti
// -- RingHom can now be applied also to BigRat
//
// Revision 1.3  2008/12/17 12:11:52  abbott
// Changed type from long to MachineInt in operations which use a machine integer
// in place of a RingElem.  The change is "superficial" but affects many files.
//
// Revision 1.2  2007/10/30 17:14:07  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.5  2007/03/07 14:07:58  bigatti
// -- minor: commented argument names for -Wextra
//
// Revision 1.4  2007/03/05 21:06:07  cocoa
// New names for homomorphism pseudo-ctors: removed the "New" prefix.
//
// Revision 1.3  2007/01/13 14:14:34  cocoa
// Overhaul of RingHom code: it nows uses SmartPtrIRC, and printing is more logical.
// Have not yet updated the documentation.
//
// Revision 1.2  2006/11/08 16:21:59  cocoa
// Structural cleaning of RingHom; many consequential changes.
//
// Revision 1.1.1.1  2006/05/30 11:39:37  cocoa
// Imported files
//
// Revision 1.3  2006/03/15 18:09:31  cocoa
// Changed names of member functions which print out their object
// into myOutputSelf -- hope this will appease the Intel C++ compiler.
//
// Revision 1.2  2006/03/12 21:28:33  cocoa
// Major check in after many changes
//
// Revision 1.1.1.1  2005/10/17 10:46:54  cocoa
// Imported files
//
// Revision 1.1.1.1  2005/05/03 15:47:31  cocoa
// Imported files
//
// Revision 1.4  2005/04/20 15:40:48  cocoa
// Major change: modified the standard way errors are to be signalled
// (now via a macro which records filename and line number).  Updated
// documentation in error.txt accordingly.
//
// Improved the documentation in matrix.txt (still more work to be done).
//
// Revision 1.3  2005/04/19 14:06:03  cocoa
// Added GPL and GFDL licence stuff.
//
// Revision 1.2  2005/02/11 14:15:20  cocoa
// New style ring elements and references to ring elements;
// I hope I have finally got it right!
//
// Revision 1.1.1.1  2005/01/27 15:12:13  cocoa
// Imported files
//
// Revision 1.5  2004/11/12 15:49:29  cocoa
// Tidying prior to 0.90 release.
// (a) documentation improved (or marked as poor)
// (b) sundry minor improvements to the code
//
// Revision 1.4  2004/07/20 15:04:06  cocoa
// The next step in the new "ring element" conversion process:
// handling the case of creating a "const RefRingElem" object
// (since C++ refuses to do this properly itself).
//
// Revision 1.3  2004/05/24 15:52:13  cocoa
// Major update:
//   new error mechanism
//   many fixes
//   RingHoms almost work now
//   RingFloat much improved
//
// Revision 1.2  2003/10/09 12:16:38  cocoa
// New coding convention for rings.
//
// Revision 1.2  2003/06/23 16:59:28  abbott
// Minor cleaning prior to public release.
//
// Revision 1.1  2003/05/14 17:16:40  abbott
// Initial revision
//
