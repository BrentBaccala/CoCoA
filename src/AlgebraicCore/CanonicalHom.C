//   Copyright (c)  2007,2009  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/CanonicalHom.H"
#include "CoCoA/RingHom.H"
#include "CoCoA/RingQQ.H"
#include "CoCoA/RingZZ.H"
#include "CoCoA/PolyRing.H"
#include "CoCoA/QuotientRing.H"
#include "CoCoA/FractionField.H"

// #include <iostream>  // for debugging only

namespace CoCoA
{

  RingHom CanonicalHom(const ring& domain, const ring& codomain)
  {
    if (domain == codomain) return IdentityHom(domain);

    // Two easy cases:
    if (IsZZ(domain)) return ZZEmbeddingHom(codomain);
    if (IsQQ(domain)) return QQEmbeddingHom(codomain); // NB result is only a partial hom!!

    // Check codomain first, as this makes it possible to exploit certain "shortcuts"
    if (IsFractionField(codomain))
    {
      if (domain == BaseRing(codomain))
        return EmbeddingHom(codomain);
      else
	return EmbeddingHom(codomain)(CanonicalHom(domain, BaseRing(codomain)));
    }
    if (IsPolyRing(codomain))
    {
      if (domain == CoeffRing(codomain))
        return CoeffEmbeddingHom(codomain);
      else
	return CoeffEmbeddingHom(codomain)(CanonicalHom(domain, CoeffRing(codomain)));
    }
    if (IsQuotientRing(codomain))
    {
      const QuotientRing QR = codomain;
      if (domain == BaseRing(QR))
        return QuotientingHom(QR);
      else
	return QuotientingHom(QR)(CanonicalHom(domain, BaseRing(QR)));
    }

    CoCoA_ERROR(ERR::CanonicalHomFail, "CanonicalHom(R1,R2)");
    return IdentityHom(codomain); // Never executed; just to keep the compiler quiet.
  }


  RingHom TmpChainCanonicalHom(const ring& domain, const ring& codomain)
  {
    //    std::clog << " TmpChainCanonicalHom ";
    try { return CanonicalHom(domain, codomain); }
    catch (const CoCoA::ErrorInfo& err) {if (err!=ERR::CanonicalHomFail) throw;}
    if (IsFractionField(codomain))
    {
      //      std::clog << " FrF ";
      return EmbeddingHom(codomain) (TmpChainCanonicalHom(domain, BaseRing(codomain)));
    }
    if (IsPolyRing(codomain))
    {
      //      std::clog << " P ";
      return CoeffEmbeddingHom(codomain) (TmpChainCanonicalHom(domain, CoeffRing(codomain)));
    }
    if (IsQuotientRing(codomain))
    {
      //      std::clog << " QR ";
      const QuotientRing QR = codomain;
      return QuotientingHom(QR) (TmpChainCanonicalHom(domain, BaseRing(QR)));
    }
    CoCoA_ERROR(ERR::CanonicalHomFail, "TmpChainCanonicalHom(R1,R2)");
    return IdentityHom(codomain); // Never executed; just to keep the compiler quiet.
  }


} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/CanonicalHom.C,v 1.10 2014/07/08 13:14:40 abbott Exp $
// $Log: CanonicalHom.C,v $
// Revision 1.10  2014/07/08 13:14:40  abbott
// Summary: Removed AsQuotientRing; added new defn of BaseRing
// Author: JAA
//
// Revision 1.9  2014/07/08 08:33:18  abbott
// Summary: Removed AsFractionField
// Author: JAA
//
// Revision 1.8  2014/07/07 12:12:17  abbott
// Summary: Removed AsPolyRing
// Author: JAA
//
// Revision 1.7  2012/02/10 10:26:40  bigatti
// -- changed RingZ.H, RingQ.H --> RingZZ.H, RingQQ.H
//
// Revision 1.6  2012/02/08 15:07:08  bigatti
// -- changed: Z,Q -> ZZ,QQ
//
// Revision 1.5  2011/02/18 12:56:08  bigatti
// -- added TmpChainCanonicalHom
//
// Revision 1.4  2009/07/24 14:21:03  abbott
// Cleaned up include directives, and added some fwd decls.
//
// Revision 1.3  2009/07/24 12:27:46  abbott
// Added some comments.
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.2  2007/03/07 11:18:01  bigatti
// -- fixed CanonicalHom goto flag
//
// Revision 1.1  2007/03/05 21:25:02  cocoa
// New CanonicalHom pseudo-ctor.
//
//
