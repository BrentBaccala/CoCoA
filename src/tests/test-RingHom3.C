//   Copyright (c)  2007  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/CanonicalHom.H"
#include "CoCoA/GlobalManager.H"
#include "CoCoA/PolyRing.H"
#include "CoCoA/QuotientRing.H"
#include "CoCoA/RingDistrMPolyClean.H"
#include "CoCoA/RingDistrMPolyInlFpPP.H"
#include "CoCoA/RingDistrMPolyInlPP.H"
#include "CoCoA/RingFp.H"
#include "CoCoA/RingHom.H"
#include "CoCoA/RingZZ.H"
#include "CoCoA/SparsePolyRing.H"
#include "CoCoA/ring.H"

using namespace CoCoA;

#include <iostream>
using std::cout;
using std::cerr;
using std::endl;
#include <vector>
using std::vector;

//----------------------------------------------------------------------
// Test for RingHom
// ZZxy->FpxII, FpxII->Fpy, FpxII->FpxII
//----------------------------------------------------------------------

#define TEST_ASSERT(cond)   do { if (!(cond)) CoCoA::AssertionFailed(#cond, __FILE__, __LINE__); } while(0)

void program()
{
  GlobalManager CoCoAFoundations;

  const QuotientRing Fp = NewZZmod(101);
  const PolyRing ZZxy = NewPolyRing(RingZZ(), 2);
  const PolyRing Fpy = NewPolyRing(Fp, 2);
  const PolyRing FpxI = NewPolyRing_DMPI(Fp, 2);
  const PolyRing FpxII = NewPolyRing_DMPII(Fp, 2);

//   RingHom ZZToFpxI = ZZEmbeddingHom(FpxI);
//   RingHom ZZToFpxII = ZZEmbeddingHom(FpxII);
  RingHom ZZToFpxI = CanonicalHom(RingZZ(), FpxI);
  RingHom ZZToFpxII = CanonicalHom(RingZZ(), FpxII);

  const vector<RingElem>& x = indets(FpxII);
  vector<RingElem> imx;
  imx.push_back(x[0]*x[0]);
  imx.push_back(x[1]*x[1]);

  const vector<RingElem>& z = indets(FpxI);
  vector<RingElem> imx1;
  imx1.push_back(z[0]*z[0]);
  imx1.push_back(z[1]*z[1]);

  const vector<RingElem>& y = indets(Fpy);
  vector<RingElem> imx2;
  imx2.push_back(y[0]*y[0]);
  imx2.push_back(y[1]*y[1]);

  RingHom ZZxyToFpxII = PolyRingHom(ZZxy, FpxII, ZZToFpxII, imx);
  RingHom FpxIIToFpxII = PolyRingHom(FpxII, FpxII, CoeffEmbeddingHom(FpxII), imx);
  RingHom FpxIIToFpy = PolyRingHom(FpxII, Fpy, CoeffEmbeddingHom(Fpy), imx2);

  RingHom ZZxyToFpxI = PolyRingHom(ZZxy, FpxI, ZZToFpxI, imx1);
//   RingHom FpxIToFpxI = PolyRingHom(FpxI, FpxI, CoeffEmbeddingHom(FpxI), imx1);
//   RingHom FpxIToFpy = PolyRingHom(FpxI, Fpy, CoeffEmbeddingHom(Fpy), imx2);
  RingHom FpxIToFpxI = PolyAlgebraHom(FpxI, FpxI, imx1);
  RingHom FpxIToFpy = PolyAlgebraHom(FpxI, Fpy, imx2);

  cout << "Simple test on polynomial ring hom" << endl
       << ZZxy << "  -->  " << FpxII << endl
       << "sending " << indet(ZZxy,0) << "  to  " << imx[0] << "  and" << endl
       << "sending " << indet(ZZxy,1) << "  to  " << imx[1] << endl;

  RingElem f = indet(ZZxy,0) + 2*indet(ZZxy,1);
  RingElem g = ZZxyToFpxII(f);
  RingElem h = ZZxyToFpxI(f);

  cout << f << " in " << owner(f) << "  maps to " << g << " in " << owner(g) << endl;
  cout << g << " in " << owner(g) << "  maps to " << FpxIIToFpxII(g) << " in " << owner(FpxIIToFpxII(g)) << endl;
  cout << g << " in " << owner(g) << "  maps to " << FpxIIToFpy(g) << " in " << owner(FpxIIToFpy(g)) << endl;

  cout << f << " in " << owner(f) << "  maps to " << h << " in " << owner(h) << endl;
  cout << h << " in " << owner(h) << "  maps to " << FpxIToFpxI(h) << " in " << owner(FpxIToFpxI(h)) << endl;
  cout << h << " in " << owner(h) << "  maps to " << FpxIToFpy(h) << " in " << owner(FpxIToFpy(h)) << endl;

  vector<RingElem> imx3;
  imx3.push_back(RingElem(Fp,13));
  imx3.push_back(RingElem(Fp,31));

  RingHom EvalFpxI = PolyAlgebraHom(FpxI, Fp, imx3);
  RingHom EvalFpxII = PolyAlgebraHom(FpxII, Fp, imx3);
  TEST_ASSERT(EvalFpxII(g) == EvalFpxI(h));
}


// Use main() to handle any uncaught exceptions and warn the user about them.
int main()
{
  try
  {
    program();
    return 0;
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    cerr << "***ERROR***  UNCAUGHT CoCoA Error";
    ANNOUNCE(cerr, err);
  }
  catch (const std::exception& exc)
  {
    cerr << "***ERROR***  UNCAUGHT std::exception: " << exc.what() << endl;
  }
  catch(...)
  {
    cerr << "***ERROR***  UNCAUGHT UNKNOWN EXCEPTION" << endl;
  }

  BuildInfo::PrintAll(cerr);
  return 1;
}
