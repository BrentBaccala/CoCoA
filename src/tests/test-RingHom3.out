Simple test on polynomial ring hom
RingWithID(3,"ZZ[x[0],x[1]]")  -->  RingWithID(6,"RingWithID(2)[x[0],x[1]]")
sending x[0]  to  x[0]^2  and
sending x[1]  to  x[1]^2
x[0] +2*x[1] in RingWithID(3,"ZZ[x[0],x[1]]")  maps to x[0]^2 +2*x[1]^2 in RingWithID(6,"RingWithID(2)[x[0],x[1]]")
x[0]^2 +2*x[1]^2 in RingWithID(6,"RingWithID(2)[x[0],x[1]]")  maps to x[0]^4 +2*x[1]^4 in RingWithID(6,"RingWithID(2)[x[0],x[1]]")
x[0]^2 +2*x[1]^2 in RingWithID(6,"RingWithID(2)[x[0],x[1]]")  maps to x[0]^4 +2*x[1]^4 in RingWithID(4,"RingWithID(2)[x[0],x[1]]")
x[0] +2*x[1] in RingWithID(3,"ZZ[x[0],x[1]]")  maps to x[0]^2 +2*x[1]^2 in RingWithID(5,"RingWithID(2)[x[0],x[1]]")
x[0]^2 +2*x[1]^2 in RingWithID(5,"RingWithID(2)[x[0],x[1]]")  maps to x[0]^4 +2*x[1]^4 in RingWithID(5,"RingWithID(2)[x[0],x[1]]")
x[0]^2 +2*x[1]^2 in RingWithID(5,"RingWithID(2)[x[0],x[1]]")  maps to x[0]^4 +2*x[1]^4 in RingWithID(4,"RingWithID(2)[x[0],x[1]]")
